tools/wptrunner/wptrunner/executors/executorwebdriver.py

```diff
@@ -646,6 +754,81 @@ def do_testharness(self, protocol, url, timeout):
 
         return rv
 
+    def bidi_deserialize(self, bidi_value):
+        """
+        Deserialize the BiDi value to the Python value, keeping non-common data typs (window) in BiDi format. The result
+        can have collisions with the classic values.
+        :param bidi_value: BiDi value to deserialize.
+        """
+        if isinstance(bidi_value, str):
+            return bidi_value
+        if isinstance(bidi_value, int):
+            return bidi_value
+        if not isinstance(bidi_value, dict):
+            raise ValueError("Unexpected bidi value: %s" % bidi_value)
+        if bidi_value["type"] == "null":
+            return None
+        if bidi_value["type"] == "boolean":
+            return bidi_value["value"]
+        if bidi_value["type"] == "number":
+            # TODO: extend with edge case values, like `Infinity`.
+            return bidi_value["value"]
+        if bidi_value["type"] == "string":
+            return bidi_value["value"]
+        if bidi_value["type"] == "array":
+            result = []
+            for item in bidi_value["value"]:
+                result.append(self.bidi_deserialize(item))
+            return result
+        if bidi_value["type"] == "object":
+            result = {}
+            for item in bidi_value["value"]:
+                result[self.bidi_deserialize(item[0])] = self.bidi_deserialize(item[1])
+            return result
+        if bidi_value["type"] == "window":
+            return bidi_value
+        # TODO: probably should return bidi value as-is, like `window` instead of raising exception. Keep for now to
+        #  check any regression in classic values.
+        raise ValueError("Unexpected bidi value: %s" % bidi_value)
+
+    def _get_next_message(self, protocol, url, test_window):
+        """
+        Get the next message from the test_driver. If the protocol supports bidi scripts, the messages are processed
+        asynchronously, otherwise the messages are processed synchronously.
+        """
+        if protocol.bidi_script:
+            # If `bidi_script` is available, use it as async call allows to process the events from the
+            # test_runner to test_driver while waiting for the next test_driver commands.
+
+            # As long as we want to be able to use scripts both in bidi and in classic mode, the script should
```

Comment:
FWIW I'd do this once when we load `self.script_resume` rather than here.

Reply:
done in https://github.com/web-platform-tests/wpt/pull/45823